src/System.Security.SecureString/src/System/Security/SafeBSTRHandle.cs

```diff
@@ -0,0 +1,83 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace System.Security
+{
+    internal sealed class SafeBSTRHandle : SafeBuffer
+    {
+        internal SafeBSTRHandle() : base(true) { }
+
+        internal static SafeBSTRHandle Allocate(uint lenInChars) => Allocate(IntPtr.Zero, lenInChars * sizeof(char));
+
+        internal static SafeBSTRHandle Allocate(IntPtr src, uint lenInBytes)
+        {
+            Debug.Assert(lenInBytes % sizeof(char) == 0);
+            SafeBSTRHandle bstr = Interop.OleAut32.SysAllocStringLen(src, lenInBytes / sizeof(char));
+            if (bstr.IsInvalid) // SysAllocStringLen returns a NULL ptr when there's insufficient memory
+            {
+                throw new OutOfMemoryException();
+            }
+            bstr.Initialize(lenInBytes);
+            return bstr;
+        }
+
+        override protected bool ReleaseHandle()
+        {
+            Interop.NtDll.ZeroMemory(handle, (UIntPtr)(Interop.OleAut32.SysStringLen(handle) * sizeof(char)));
+            Interop.OleAut32.SysFreeString(handle);
+            return true;
+        }
+
+        internal unsafe void ClearBuffer()
+        {
+            byte* bufferPtr = null;
+            try
+            {
+                AcquirePointer(ref bufferPtr);
+                Interop.NtDll.ZeroMemory((IntPtr)bufferPtr, (UIntPtr)(Interop.OleAut32.SysStringLen((IntPtr)bufferPtr) * sizeof(char)));
+            }
+            finally
+            {
+                if (bufferPtr != null)
+                {
+                    ReleasePointer();
+                }
+            }
+        }
+
+        internal unsafe uint Length => Interop.OleAut32.SysStringLen(this);
+
+        internal unsafe static void Copy(SafeBSTRHandle source, SafeBSTRHandle target, uint bytesToCopy)
+        {
+            if (bytesToCopy == 0)
+            {
+                return;
+            }
+
+            byte* sourcePtr = null, targetPtr = null;
+            try
+            {
+                source.AcquirePointer(ref sourcePtr);
+                target.AcquirePointer(ref targetPtr);
+
+                Debug.Assert(Interop.OleAut32.SysStringLen((IntPtr)sourcePtr) * sizeof(char) >= bytesToCopy, "Source buffer is too small.");
```

Comment:
> Previously it used Contract.Assert. Is that no longer the pattern?

Across corefx we've settled on using Debug.Assert instead of Contract.Assert.

> is it worth having a retail exception (that should be unreachable) just before any memcpy?

We effectively do: MemoryCopy does such checks.  That's why it takes the destination size as the third argument; its only use in MemoryCopy is in validating that the fourth argument is within range.  That will ensure we're not writing to any memory we don't own.  This Assert is validating that we're not reading from any memory we don't own.  I personally think it'd be overkill to add an exception-throwing check here for the source.